### PR TITLE
ECS simple implementation

### PR DESCRIPTION
**Pull request by GuanHuiLim**
Creation and deletion of entities & components
Registering and usage of systems
Views for iteration through entities with specified components

### CODE REVIEW

File: Engine/src/Engine/ECS/ECS.cpp [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include "pch.h"
+#include "ECS.h"
```
Comment (spironan) [inline Engine/src/Engine/ECS/ECS.cpp:2]: remove this file if its not used.

File: Engine/src/Engine/ECS/ECS.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+#include "pch.h"
+#include "ECS_Utility.h"
+#include "SparseArray.h"
```
Comment (spironan) [inline Engine/src/Engine/ECS/ECS.h:4]: same as before, should remove file if not used

File: Engine/src/Engine/ECS/ECS_Test.cpp [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#include "pch.h"
+#include "ECS_Test.h"
```
Comment (spironan) [inline Engine/src/Engine/ECS/ECS_Test.cpp:2]: once again same as before should remove if unused

File: Engine/src/Engine/ECS/SparseArray.h [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+namespace engine
+{
+    template<typename T, std::size_t MAX_SIZE, typename DenseAllocator = std::allocator<T>, 
+        typename SparseAllocator = std::allocator<std::size_t>>
+    class SparseContainer
+    {
+
+
+    private:
+        /*using alloc_type = typename std::allocator_traits<Allocator>::template rebind_alloc<T>;
```
Comment (spironan) [inline Engine/src/Engine/ECS/SparseArray.h:15]: Delete away if unused?

File: Engine/src/Engine/ECS/System.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+Copyright (C) 2021 DigiPen Institute of Technology.
+Reproduction or disclosure of this file or its contents
+without the prior written consent of DigiPen Institute of
+Technology is prohibited.
+ *********************************************************************/
+#pragma once
+#include <set>
+#include "ECS_Utility.h"
+class System
```
Comment (spironan) [inline Engine/src/Engine/ECS/System.h:18]: Should Systems be wrapped in engine namespace or not?

File: Engine/src/Engine/ECS/SystemManager.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+Technology is prohibited.
+ *********************************************************************/
+#pragma once
+
+#include <cassert>
+#include <memory>
+#include <unordered_map>
+#include "ECS_Utility.h"
+#include "System.h"
+class SystemManager
```
Comment (spironan) [inline Engine/src/Engine/ECS/SystemManager.h:22]: similarly for system manager

**Review by spironan — CHANGES_REQUESTED** [review]
Do Get back on the comments before approval

File: Engine/src/Engine/ECS/ECS.cpp (hunk 1 above)
Comment (GuanHuiLim) [reply Engine/src/Engine/ECS/ECS.cpp:2]: It has been removed

File: Engine/src/Engine/ECS/ECS.h (hunk 2 above)
Comment (GuanHuiLim) [reply Engine/src/Engine/ECS/ECS.h:4]: This is the main header file used by others to access the ecs functionality

File: Engine/src/Engine/ECS/ECS_Test.cpp (hunk 3 above)
Comment (GuanHuiLim) [reply Engine/src/Engine/ECS/ECS_Test.cpp]: It has been removed

File: Engine/src/Engine/ECS/SparseArray.h (hunk 4 above)
Comment (GuanHuiLim) [reply Engine/src/Engine/ECS/SparseArray.h:15]: In the future, this class will likely be improved upon, especially in terms of memory allocation. So thats why it is there

File: Engine/src/Engine/ECS/System.h (hunk 5 above)
Comment (GuanHuiLim) [reply Engine/src/Engine/ECS/System.h]: It has been wrapped

File: Engine/src/Engine/ECS/SystemManager.h (hunk 6 above)
Comment (GuanHuiLim) [reply Engine/src/Engine/ECS/SystemManager.h]: It has been wrapped

File: Engine/src/Engine/ECS/SparseArray.h (hunk 4 above)
Comment (spironan) [reply Engine/src/Engine/ECS/SparseArray.h:15]: ah ok

File: Engine/src/Engine/ECS/ECS.h (hunk 2 above)
Comment (spironan) [reply Engine/src/Engine/ECS/ECS.h:4]: ohhh ok then just put a note that we just need to include this file

**Review by spironan — APPROVED** [review]
ok looks good besides the lack of comments

**Pull request merged**